x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/PersistentCache.java

```diff
@@ -0,0 +1,623 @@
+     * matching docs in the map of Lucene documents) then the file is deleted from disk. If a doc is found the stored fields are extracted
+     * from the Lucene document and are used to rebuild the necessary {@link CacheKey}, {@link SnapshotId}, {@link IndexId}, {@link ShardId}
+     * and cache file ranges objects. The Lucene document is then indexed again in the new persistent cache index (the current
+     * {@link CacheIndexWriter}) and the cache file is added back to the searchable snapshots cache again. Note that adding cache
+     * file to the cache service might trigger evictions so previously reindexed Lucene cache files might be delete again (see
+     * CacheService#onCacheFileRemoval(CacheFile) method which calls {@link #removeCacheFile(CacheFile)}.
+     *
+     * @param cacheService the {@link CacheService} to use when repopulating {@link CacheFile}.
+     */
+    void loadCacheFiles(CacheService cacheService) {
+        ensureOpen();
+        try {
+            for (CacheIndexWriter writer : writers) {
+                final NodeEnvironment.NodePath nodePath = writer.nodePath();
+                logger.debug("loading persistent cache on data path [{}]", nodePath);
+
+                for (String indexUUID : nodeEnvironment.availableIndexFoldersForPath(nodePath)) {
+                    for (ShardId shardId : nodeEnvironment.findAllShardIds(new Index("_unknown_", indexUUID))) {
+                        final Path shardDataPath = writer.nodePath().resolve(shardId);
+                        final Path shardCachePath = getShardCachePath(new ShardPath(false, shardDataPath, shardDataPath, shardId));
+
+                        if (Files.isDirectory(shardCachePath)) {
+                            logger.trace("found snapshot cache dir at [{}], loading cache files from disk and index", shardCachePath);
+                            Files.walkFileTree(shardCachePath, new CacheFileVisitor(cacheService, writer, documents));
+                        }
+                    }
+                }
+            }
+            for (CacheIndexWriter writer : writers) {
+                writer.prepareCommit();
+            }
+            for (CacheIndexWriter writer : writers) {
+                writer.commit();
+            }
+            logger.info("persistent cache index loaded");
+            documents.clear();
+        } catch (IOException e) {
+            try {
+                close();
+            } catch (Exception e2) {
+                logger.warn("failed to close persistent cache index", e2);
+                e.addSuppressed(e2);
+            }
+            throw new UncheckedIOException("Failed to load persistent cache", e);
+        } finally {
+            closeIfAnyIndexWriterHasTragedyOrIsClosed();
+        }
+    }
+
+    void commit() throws IOException {
+        ensureOpen();
+        try {
+            for (CacheIndexWriter writer : writers) {
+                writer.prepareCommit();
```

Comment:
Do we actually need two loops for preparing and then committing? We do that for the CS persistence so we can throw `IOError` and stop everything if a commit fails but for the cache service it seems like we could just `commit` in a loop and ahve that deal with `prepareCommit`? That would also make exceptions less noisy because if you prepare a commit on one writer  and then fail committing another writer and thus call close `close` the first writer will throw on closing because it has a pending commit as well.

Reply:
That's a good point, thanks for catching this. Commiting in a single loop makes it less error prone as you pointed. I pushed ac6050f206ff4e096736ef19a2812c33a365b9b7